### DISCUSSION TITLE
Validation only password and confirmation and there was an error

### DISCUSSION THREAD

**Discussion by jlopes90** [codeigniter4/shield#359] [Q&A] [answered]
### PHP Version

7.4.30

### CodeIgniter4 Version

4.2.3

### Shield Version

develop

### Which operating systems have you tested for this bug?

Windows

### Which server did you use?

apache

### Database

MySQL 5.7.36

### Did you customize Shield?

~~No~~ Yes.

### What happened?

I filled in the form with password and confirm password and validation failed.

### Steps to Reproduce

```php
public function resetAction()
{
    $rules = [
        'password' => [
            'required',
            'strong_password'
        ],
        'password_confirm' => [
            'required',
            'matches[password]',
        ],
    ];
    
    if (!$this->validate($rules)) {
        return redirect()->back()->withInput()->with('errors', $this->validator->getErrors());
    }

    {...}
}
```

![imagem](https://user-images.githubusercontent.com/50026015/183886080-61d1ee7f-34e3-4c8e-af9c-af1eca6c7c34.png)

### Expected Output

![imagem](https://user-images.githubusercontent.com/50026015/183885527-e3d6b791-672c-4c86-a5dc-ac0b3ca2b959.png)

**Comment by kenjis** [comment]
I think Shied does not provide Rest Password page.

**Comment by datamweb** [comment]
Hey @jlopes90 , Thanks for your report,
The page you provided is a custom page with custom code.
Therefore, we cannot consider the issue as a bug.
Maybe someone can help you by providing more details of your custom code in the discussion  section.

Do you agree with me to move this topic to the discussion section?

**Comment by jlopes90** [comment]
> Do you agree with me to move this topic to the discussion section?

yes

**Comment by jlopes90** [comment]
Because `strong_password` gave error, without this it works.

**Reply by datamweb** [reply]
I suggest you provide more details about the controller, view or any other file that you have changed.

**Comment by jlopes90** [comment]
The fields return `username`, `email` and `password` and the `email` post is an empty string so user forces the `setEmail()` call which gives an error.

https://github.com/codeigniter4/shield/blob/a76b35e7bf6746037b94061fab08cfcd08ac89b5/src/Authentication/Passwords/ValidationRules.php#L56-L69

```php
protected function buildUserFromRequest(): User 
{ 
    $fields = $this->prepareValidFields(); 
    
    /** @var IncomingRequest $request */ 
    $request = service('request'); 
    
    $data = $request->getPost($fields);
    
    if (empty($data['username'])){
      unset($data['username']);
    }
    if (empty($data['email'])){
      unset($data['email']);
    }
    
    return new User($data);
}
```

---

Forces to validate email is empty string, the result gives an error, I used to check if email is empty, just ignore it.

https://github.com/codeigniter4/shield/blob/a76b35e7bf6746037b94061fab08cfcd08ac89b5/src/Authentication/Passwords/NothingPersonalValidator.php#L68-L87

```php
// Parse out as many pieces as possible from username, password and email.
// Use the pieces as needles and haystacks and look every which way for matches.
if ($valid) {
    // Take username apart for use as search needles
    $needles = $this->strip_explode($userName);

    // If you don't use email
    if (!empty($email)) {

        // extract local-part and domain parts from email as separate needles
        [
            $localPart,
            $domain,
        ] = explode('@', $email);
        // might be john.doe@example.com and we want all the needles we can get
        $emailParts = $this->strip_explode($localPart);
        if (!empty($domain)) {
            $emailParts[] = $domain;
        }
        $needles = array_merge($needles, $emailParts);
    }

    // Get any other "personal" fields defined in config
    $personalFields = $this->config->personalFields;

    {...}
}
```

---

I edited it but I know it's not a good idea.

**Comment by kenjis** [comment]
Add email address to the data to validate.
`strong_password` needs email address to check.

**Comment by jlopes90** [comment, marked as answer]
I did like this:

1. Forgot password is sent.
2. Receive email and open link `.../reset?email=xxx&token=xxx`.
3. The password reset form appears, the e-mail and the token are hidden in the input and fill in password and confirm.
4. And save change and that's it.
